crates/sui-faucet/src/faucet/simple_faucet.rs

```diff
@@ -46,132 +51,162 @@ impl SimpleFaucet {
+            if let Err(e) = producer.send(*coin.id()).await {
+                panic!("Failed to set up gas pools: {:?}", e);
+            }
         }
 
-        let primary_coin = &coins[coins.len() - 1];
-        let gas_coin = &coins[coins.len() - 2];
-
-        info!(
-            "Using {} as primary, {} as the gas payment",
-            primary_coin, gas_coin
-        );
+        info!("Using coins: {:?}", coins);
 
         Ok(Self {
             wallet,
-            primary_coin_id: *primary_coin.id(),
-            gas_coin_id: *gas_coin.id(),
             active_address,
+            producer: Mutex::new(producer),
+            consumer: Mutex::new(consumer),
         })
     }
 
-    async fn get_coins(
-        &self,
-        amounts: &[u64],
-    ) -> Result<(Vec<SuiParsedObject>, TransactionDigest), FaucetError> {
-        let (coins, tx_digest) = self
-            .split_coins(
-                amounts,
-                self.primary_coin_id,
-                self.gas_coin_id,
-                self.active_address,
-                DEFAULT_GAS_BUDGET,
-            )
-            .await
-            .map_err(|err| FaucetError::Wallet(err.to_string()))?;
-
-        Ok((coins, tx_digest))
+    async fn select_coins(&self, number_of_coins: usize) -> Vec<ObjectID> {
+        assert!(number_of_coins > 0);
+        // If the gas candidate queue is exhausted, the request will be
+        // suspended indefinitely until a producer puts in more candidate
+        // gas objects. At the same time, other requests will be blocked by the
+        // lock acquisition as well.
+        let mut consumer = self.consumer.lock().await;
+        let mut coins = Vec::with_capacity(number_of_coins);
+        while let Some(coin) = consumer.recv().await {
+            // TODO: for now we assume each SUI object is enough to cover the split
+            // but this may not be true, if we run the faucet for really really long time or
+            // due to some other unexpected issues.
+            coins.push(coin);
+            if coins.len() == number_of_coins {
+                break;
+            }
+        }
+        debug!("Planning to use coins: {:?}", coins);
+        coins
     }
 
-    async fn public_transfer_objects(
+    async fn transfer_gases(
         &self,
-        coins: &[ObjectID],
-        recipient: SuiAddress,
-    ) -> Result<Vec<TransactionDigest>, FaucetError> {
-        let mut digests = Vec::with_capacity(coins.len());
-        for coin_id in coins.iter() {
-            let digest = self
-                .public_transfer_object(
+        amounts: &[u64],
+        to: SuiAddress,
+    ) -> Result<Vec<(TransactionDigest, ObjectID, u64, ObjectID)>, FaucetError> {
+        let number_of_coins = amounts.len();
+        let coins = self.select_coins(number_of_coins).await;
+        let futures: Vec<_> = coins
+            .iter()
+            .zip(amounts)
+            .map(|(coin_id, amount)| {
+                self.transfer_sui(
```

Comment:
As a further improvement, we can construct a batch transaction instead of having 5 single transactions. No need to do this in this PR

Reply:
I indeed tried to do it in this PR! but I decided to gave up because batch txn isn't the easier to construct compared to single txn at the moment! Agree we can revisit this if desired.